### PR TITLE
fix Barline change issues

### PR DESCRIPTION
**Pull request by sammik**
Resolves: #19782 
Resolves: #19778 

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [ ] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

**Comment by zacjansheski** [comment]
With regard to https://github.com/musescore/MuseScore/issues/19776
1. Add forward repeat
2. "M" for multi-measure rest
3. Change forward repeat to double barline
4. "M" to remove multi-measure rest
5. Change double barline to single barline
6. "M" for multi-measure rest (see artifact)
7. Undo all
8. Redo all, crash (After change forward repeat to double barline)

Not necessarily a blocker, so far this PR is a big improvement. :)

https://github.com/musescore/MuseScore/assets/69917893/cb519408-3ca9-4129-8b49-f500a5440d58

**Comment by oktophonie** [comment]
Re. #19782 point 2, it seems that deleting the end repeat at a system break when in MM-rest mode is impossible, though deleting the start repeat works fine)

**Comment by zacjansheski** [comment]
Adding/deleting the repeats attached to mm rests is possible in both nightly and PR, but they don't appear until the mmrest is removed.

As @oktophonie said, the initial repeat works as desired but the end repeat does not.
https://github.com/musescore/MuseScore/assets/69917893/d7a3f308-95d7-4244-b97c-ff898d62a392

**Comment by sammik** [comment]
@oktophonie @zacjansheski 
Repeat barline should be fixed now.

**Comment by zacjansheski** [comment]
Tested on Mac OS 14, Ubuntu 20.04.6, Windows 11.  Approved
https://github.com/musescore/MuseScore/issues/19782 - FIXED
https://github.com/musescore/MuseScore/issues/19778 - FIXED
https://github.com/musescore/MuseScore/issues/19776 - FIXED

File: src/engraving/dom/engravingobject.cpp [hunk 1]
```diff
@@ -475,6 +475,12 @@ void EngravingObject::undoChangeProperty(Pid id, const PropertyValue& v, Propert
                 toEngravingItem(this)->manageExclusionFromParts(v.toBool());
             }
         }
+    } else if (id == Pid::REPEAT_START || id == Pid::REPEAT_END) {
+        // change also coresponding mmRestMeasure
+        Measure* m = const_cast<Measure*>(toMeasure(this)->coveringMMRestOrThis());
```
Comment (Eism) [inline src/engraving/dom/engravingobject.cpp:480]: Handling specific logic (child logic) in parent classes is a bad idea. 
You should override the `undoChangeProperty` method and process this logic where it's necessary.

File: src/engraving/dom/measure.h [hunk 2]
```diff
@@ -300,6 +300,10 @@ class Measure final : public MeasureBase
     bool setProperty(Pid propertyId, const PropertyValue&) override;
     PropertyValue propertyDefault(Pid) const override;
 
+    void undoChangeProperty(Pid id, const PropertyValue& newValue);
+    void undoChangeProperty(Pid id, const PropertyValue& newValue, PropertyFlags ps) override;
+    using EngravingObject::undoChangeProperty;
```
Comment (Jojo-Schmitz) [inline src/engraving/dom/measure.h:305]: is that `using` needed?

Comment (sammik) [reply src/engraving/dom/measure.h]: It seems, it is not. Thanks!

File: src/engraving/dom/engravingobject.cpp (hunk 1 above)
Comment (sammik) [reply src/engraving/dom/engravingobject.cpp]: @Eism done!

**Review by Eism — APPROVED** [review]

**Comment by zacjansheski** [comment]
Tested on MacOS 14, Windows 11, Ubuntu 22.04.3.  Approved

https://github.com/musescore/MuseScore/issues/19782 FIXED
https://github.com/musescore/MuseScore/issues/19778 FIXED

https://github.com/musescore/MuseScore/issues/19776 - We should keep this open.  Still able to crash MuseScore with the described steps
[UndoRedoCrash.zip](https://github.com/musescore/MuseScore/files/13217101/UndoRedoCrash.zip)

**Comment by Eism** [comment]
Unlinked https://github.com/musescore/MuseScore/issues/19776 and merged

**Pull request merged**